keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -613,7 +621,178 @@ sequence that it will not successfully decode.
+
+1. If both Raw and Object are `nil`, `null` is returned.
+1. If Raw is not `nil`, return it verbatim.
+1. Otherwise (Raw is `nil` and Object is not `nil`), return the result of marshalling Object.
+
+Note that, in the second case, the bytes of the Raw field must be a valid JSON text in order to
+successfully serialize an object containing a `RawExtension` to JSON.
+
+#### Usage
+
+##### Transient External Types
+
+External versioned types may use `RawExtension` to exchange arbitrary objects and plugins without
+persisting them to storage. In these cases, only a single object encoding is involved. When
+preparing to send, or handle a received object containing `RawExtension`, callers can assume that
+the Raw bytes are in the same encoding as the negotiated request or response encoding.
+
+##### Stored External Types
+
+Storing the verbatim Raw bytes of a `RawExtension` received from a client introduces additional
+considerations on top of the transient (transmit-only) case. The encoding of the Raw bytes is
+determined by encoding of the request that wrote the value of the RawExtension, which may or may not
+be the same as the object's storage encoding.
+
+##### Types as Canonical Definition of Custom Resources
+
+Throughout the ecosystem, it is common practice to maintain Go structs as the canonical definition
+for API extensions. In many cases, `controller-gen` is used to mechanically translate such types
+from Go sources to CustomResourceDefinition manifests. Similarly, `client-gen` can produce typed Go
+clients that use the canonical Go types directly. These Go struct types can and sometimes do include
+fields of type `RawExtension`
+([example](https://github.com/openshift/api/blob/944467d2cc3b03225ccc24c4e88b876396202d5a/operator/v1/types.go#L91)).
+
+#### Scenarios
+
+The following table enumerates API request and response flows that can involve `RawExtension`.
+
+The *Client* and *Server* columns indicate the types the named component uses to processes API
+objects. If "dynamic", it uses Unstructured (e.g. a custom resource handler or a dynamic client). If
+"static", it uses API-specific Go types that may include `RawExtension` (e.g. clients generated by
+`client-gen`, kube-apiserver built-in types, aggregated apiservers). The table omits cases where
+both the client and the server are dynamic (e.g. a dynamic client and a custom resource handler),
+since neither side should be dealing with `RawExtension` values. The edge case where a client
+program makes a `RawExtension` a child of an Unstructured value's `map[string]interface{}` can be
+considered a static client case for the purposes of this evaluation.
+
+The *Encoding* column is the client's encoding of the request body (for requests) or the server's
+encoding of the response body (for responses).
+
+| N  | Client  | Server  | Direction | Encoding |
+|----|---------|---------|-----------|----------|
+| 1  | static  | static  | request   | json     |
```

Comment:
supernit: I'm used to the client for the "static" case as being referred to as a "typed" client.

Reply:
:+1: